run.sh

```
@@ -336,28 +336,107 @@ open_temporary_ngrok() {
 
+kill_stale_tailscale_funnels() {
+    # Stop leftover `tailscale funnel --bg=false <port>` processes. They are
+    # often root-owned (started with sudo), so escalate when a plain kill fails.
+    local pids remaining
+    pids="$(pgrep -f "tailscale[[:space:]]+funnel[[:space:]]+--bg=false[[:space:]]+${NGROK_PORT}" 2>/dev/null || true)"
+    [ -n "$pids" ] || return 0
+    warn "Stopping stale Tailscale Funnel processes for port $NGROK_PORT."
+    if [ "$(id -u)" = "0" ]; then
+        printf '%s\n' "$pids" | xargs kill 2>/dev/null || true
+    else
+        sudo kill $pids 2>/dev/null || printf '%s\n' "$pids" | xargs kill 2>/dev/null || true
```

Shellcheck (0.11.0)

[info] Double quote to prevent globbing and word splitting.

(SC2086)